### PR TITLE
Tool concurrency classification (is_mutating)

### LINKED ISSUE #937: Tool concurrency classification (is_mutating)
## Problem

LoongClaw's tool system has no concurrency classification. When multiple tool calls could run in parallel, the runtime has no way to know which tools are safe to batch (read-only) vs which must be serialized (mutating). This forces conservative sequential execution.

Competitors handle this:
- **Codex** has `ToolHandler::is_mutating()` — a per-invocation trait method that returns whether this specific call mutates state (dynamic, because some tools are read-only with certain args and mutating with others)
- **Nanobot** uses `concurrency_safe` / `exclusive` partitioning on Tool ABC — tools declare themselves as batchable or exclusive
- **FastClaw** maintains a `readOnlyTools` map — static declaration at registration time

## What To Build

Add a `ConcurrencyClass` to tool metadata:

```rust
enum ConcurrencyClass {
    ReadOnly,   // safe to run in parallel with other ReadOnly tools
    Mutating,   // must be serialized
    Unknown,    // default, treated as Mutating for safety
}
```

### Integration points
- Add the enum to `ToolDescriptor` or equivalent in the tool metadata
- Add a method to `CoreToolAdapter` trait (or keep it static on the descriptor)
- Annotate built-in tools with correct classification:
  - ReadOnly: file read, grep, glob, ls, web search
  - Mutating: file write, edit, bash, git operations
  - Unknown: default for extension/MCP tools

### Static vs Dynamic
The RFC's open question: static (declared in descriptor) or dynamic (per-invocation `is_mutating()` method)? Codex uses dynamic because some tools' mutability depends on arguments (e.g., a `git` tool that can be `git status` or `git push`). Recommend starting with static classification and adding dynamic override later if needed.

## Target Files

- `crates/app/src/tools/catalog.rs` — tool catalog/registry
- `crates/kernel/src/tool.rs` — `CoreToolAdapter` trait

## Scope

~50 LoC. This is a Phase 1 quick win — smallest scope item.

## Acceptance Criteria

- [ ] `ConcurrencyClass` enum defined
- [ ] Added to tool descriptor/metadata
- [ ] Built-in tools annotated with correct classification
- [ ] Default `Unknown` treated as `Mutating`
- [ ] `task verify` passes in loongclaw-dev

## References

- [EPIC-001 RFC — Q2: Tool/Skill System](loongclaw-cases/cases/CASE-20260405-competitive-analysis/research/rfc.md)
- Part of epic #976

### PR DESCRIPTION
**Pull request by gh-xj**
## Summary

- Problem: the repo exposed scheduling metadata (`ToolSchedulingClass`) but not an explicit concurrency classification surface at the kernel/tool-catalog layer, which is what issue `#937` asked to add.
- Why it matters: reviewers and downstream consumers need a stable concurrency signal without inferring semantics from scheduler internals, and the metadata needs to stay conservative for wrapper and serial read-only tools.
- What changed: added `ToolConcurrencyClass` to the kernel surface, removed the misleading adapter-level hook, exposed `concurrency_class` in the app tool catalog, classified built-in tools conservatively including feature-gated feishu tools, and added focused regression coverage for both rust-level and serialized catalog metadata.
- What did not change (scope boundary): did not rewrite batch execution logic, did not rename `ToolSchedulingClass`, and did not change tool scheduling or approval behavior.

## Linked Issues

- Closes #937
- Related #976

## Change Type

- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [ ] Documentation
- [ ] Security hardening
- [ ] CI / workflow / release

## Touched Areas

- [x] Kernel / policy / approvals
- [ ] Contracts / protocol / spec
- [ ] Daemon / CLI / install
- [ ] Providers / routing
- [x] Tools
- [ ] Browser automation
- [ ] Channels / integrations
- [ ] ACP / conversation / session runtime
- [ ] Memory / context assembly
- [ ] Config / migration / onboarding
- [ ] Docs / contributor workflow
- [ ] CI / release / workflows

## Risk Track

- [x] Track A (routine / low-risk)
- [ ] Track B (higher-risk / policy-impacting)

## Validation

- [ ] `cargo fmt --all -- --check`
- [ ] `cargo clippy --workspace --all-targets --all-features -- -D warnings`
- [ ] `cargo test --workspace --locked`
- [ ] `cargo test --workspace --all-features --locked`
- [ ] Relevant architecture / dep-graph / docs checks for touched areas
- [x] Additional scenario, benchmark, or manual checks when behavior changed
- [ ] If this changes config/env fallback, limits, or defaults: include before/after behavior and regression coverage for explicit path, fallback path, and boundary values
- [ ] If tests mutate process-global env: document how state is restored or serialized

Commands and evidence:

```text
cargo test -p loongclaw-kernel core_tool_adapter_defaults_to_unknown_concurrency_class -- --exact
  PASS (1 passed)

cargo test -p loongclaw-app tool_catalog_entries_expose_concurrency_class
  PASS (1 passed)

cargo test -p loongclaw-app scheduling_class_marks_parallel_safe_subset
  PASS (1 passed)

cargo test -p loongclaw-app autonomy_capability_action_catalog_entries_expose_serializable_metadata
  PASS (1 passed)

cargo test -p loongclaw-app --all-features --lib feishu_tool_catalog_entries_expose_explicit_concurrency_class -- --nocapture
  PASS (1 passed)

cargo test -p loongclaw-app --all-features --lib feishu_resource_download_catalog_entry_is_mutating -- --nocapture
  PASS (1 passed)

cargo clippy -p loongclaw-app -p loongclaw-kernel --all-targets --all-features -- -D warnings
  PASS
```

## User-visible / Operator-visible Changes

- Tool metadata now exposes an explicit `concurrency_class` field (`read_only`, `mutating`, or `unknown`) alongside existing scheduling metadata, including feature-gated feishu tools in all-features builds.

## Failure Recovery

- Fast rollback or disable path: revert commits `ca7c047e`, `b9e77a24`, `fec81136`, and `bb2eb6e4` from the branch.
- Observable failure symptoms reviewers should watch for: incorrect `concurrency_class` values in serialized tool catalog output, especially for wrapper tools like `tool.invoke` or serial read-only tools such as `session_status`.

## Reviewer Focus

- `crates/kernel/src/tool.rs`: confirm the new kernel surface is appropriately conservative and defaults to `unknown`.
- `crates/app/src/tools/catalog.rs`: confirm the mapping stays `ParallelSafe -> read_only` and `SerialOnly -> unknown`, and that no scheduling behavior changed.
- `crates/app/src/tools/catalog.rs`: review the serialized catalog regression coverage for the new `concurrency_class` field.

## Summary by CodeRabbit

* **New Features**
  * Catalog now includes concurrency classification for tools (read-only, mutating, unknown), surfaced in serialized metadata to inform execution behavior.
  * Integration tools (including Feishu) now expose non-unknown concurrency information where applicable.

* **Tests / Quality**
  * Added coverage to verify concurrency values appear correctly in the catalog and integrations, improving reliability.

### CODE REVIEW

File: crates/kernel/src/tool.rs [hunk 1]
```diff
@@ -9,10 +10,32 @@ pub use loongclaw_contracts::{
...
+    }
+}
+
 #[async_trait]
 pub trait CoreToolAdapter: Send + Sync {
     fn name(&self) -> &str;
 
+    fn concurrency_class(&self) -> ToolConcurrencyClass {
+        ToolConcurrencyClass::Unknown
+    }
```
[the remark this answers is not in the document]
Comment (gh-xj) [reply crates/kernel/src/tool.rs]: Fixed in `5380dcb2`.

The adapter-level `CoreToolAdapter::concurrency_class()` hook was removed instead of being kept as a misleading per-adapter API. In this repo the main `MvpToolAdapter` multiplexes many tools by `ToolCoreRequest.tool_name`, so concurrency classification now lives only on per-tool descriptor metadata.

The kernel side now encodes the conservative rule via `ToolConcurrencyClass::requires_serial_execution()`, with `Unknown` treated the same as `Mutating` for execution safety.

Verification run on the updated head:
- `cargo test -p loongclaw-kernel unknown_concurrency_class_requires_serial_execution`
- `cargo clippy -p loongclaw-kernel -p loongclaw-app --all-targets --all-features -- -D warnings`
- `cargo fmt --all -- --check`

File: crates/app/src/tools/catalog.rs [hunk 2]
```diff
@@ -293,6 +294,13 @@ impl ToolDescriptor {
         self.policy.scheduling_class
     }
 
+    pub fn concurrency_class(&self) -> ToolConcurrencyClass {
+        match self.scheduling_class() {
+            ToolSchedulingClass::ParallelSafe => ToolConcurrencyClass::ReadOnly,
+            ToolSchedulingClass::SerialOnly => ToolConcurrencyClass::Unknown,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (gh-xj) [reply crates/app/src/tools/catalog.rs:300]: Fixed in `5380dcb2`.

`ToolDescriptor` now owns explicit `concurrency_class` metadata instead of deriving it from `scheduling_class`. Known mutators such as `delegate_async`, `file.write`, `file.edit`, `shell.exec`, and `bash.exec` now emit `mutating`, while wrappers like `tool.invoke` remain `unknown`.

Verification run on the updated head:
- `cargo test -p loongclaw-app tool_catalog_entries_expose_concurrency_class`
- `cargo test -p loongclaw-app autonomy_capability_action_catalog_entries_expose_serializable_metadata`
- `cargo test -p loongclaw-app scheduling_class_marks_parallel_safe_subset`

**Review by chumyin — APPROVED** [review]
hey @gh-xj, thanks for pushing this through — i tightened the feature-gated feishu classifications so the built-in catalog stays explicit under all-features too, added regression coverage for that path, and reran the relevant checks. this is in a really solid place now, nice work.
